File: Algo201/problem_of_the_class/java/Zuma.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class Zuma {
    private static final BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

    public static void main(String[] args) throws IOException {
        String[] input = in.readLine().split(" ");
        System.out.println(findMinStep(input[0], input[1]));
        in.close();
    }
    
    public static int findMinStep(String board, String hand) {
        if (board == null) {
            return 0;
        }

        HashMap<Character, Integer> hitAmount = new HashMap<Character, Integer>();
        for (int i = 0; i < hand.length(); i++) {
            char c = hand.charAt(i);
            hitAmount.put(c, hitAmount.getOrDefault(c, 0) + 1);
        }

        return DepthFirstSearch(board, hitAmount, new HashMap<String, Integer>());
    }

    public static int DepthFirstSearch(String board, Map<Character, Integer> hitAmount, Map<String, Integer> temp) {
        if (temp.containsKey(board)) {
            return temp.get(board);
        }

        if (board.isEmpty())
            return 0;

        int i = 0;
        int ans = Integer.MAX_VALUE;
        StringBuilder currentBoard = new StringBuilder(board);
        while (i < currentBoard.length()) {
            for (char c : hitAmount.keySet()) {
                if (hitAmount.get(c) != null && hitAmount.get(c) > 0) {
                    currentBoard.insert(i + 1, c);
                    hitAmount.put(c, hitAmount.get(c) - 1);

                    String newBoard = updateBoard(currentBoard.toString());
                    int r = DepthFirstSearch(newBoard, hitAmount, temp);
                    if (r != -1) {
                        ans = Math.min(ans, r + 1);
                    }

                    // backtracking
                    currentBoard.deleteCharAt(i + 1);
                    hitAmount.put(c, hitAmount.get(c) + 1);
                }
            }

            i++;
        }
        ans = ans == Integer.MAX_VALUE ? -1 : ans;
        temp.put(board, ans);
        return ans;
    }

    public static  String updateBoard(String board) {
        int i = 0;
        while (i < board.length()) {
            int j = i;
            while (j < board.length() && (board.charAt(i) == board.charAt(j)))
                j++;
            if (j - i >= 3) {
                board = board.substring(0, i) + board.substring(j);
                return updateBoard(board);
            } else {
                i++;
            }
        }
        return board;
    }
}

File: Algo201/problem_of_the_class/python/zuma.py
import functools

class Solution:
    def findMinStep(self, board: str, hand: str) -> int:
        self.min_moves = float("inf")

        @functools.lru_cache(None)
        def zuma_hit(board: tuple) -> tuple:
            ball = []
            N = len(board)
            i = 0
            hit_again = False
            while i < N:
                j = i
                while j < N and board[j] == board[i]:
                    j += 1
                if j - i <= 2:
                    ball.extend(board[i:j])
                else:
                    hit_again = True
                i = j
            ball = tuple(ball)
            if hit_again:
                ball = zuma_hit(ball)
            return ball

        @functools.lru_cache(None)
        def depth_first_seach(board: tuple, hand: tuple, moves: int):
            hand = tuple(b for b in hand if b in board and (hand.count(b) + board.count(b)) >= 3)

            board = zuma_hit(board)

            if len(board) == 0:
                self.min_moves = min(self.min_moves, moves)
            elif len(hand) <= 0:
                return
            else:
                n = 0
                while n < len(hand):
                    zuma_ball = list(hand)
                    ball = zuma_ball.pop(n)

                    steps = set()

                    i = 0
                    count = 0
                    for j in range(len(board)):
                        if board[i] == board[j]:
                            count += 1
                        else:
                            i = j
                            count = 1
                        if count == 1 and board[i] == ball:
                            steps.add((i, ball))

                    for j in range(1, len(board)):
                        if j < (len(board) - 1):
                            if board[j-1] != ball and board[j+1] != ball:
                                steps.add((j, ball))

                    for i, k in steps:
                        after_hit = list(board)
                        after_hit.insert(i, k)
                        depth_first_seach(tuple(after_hit), tuple(zuma_ball), moves + 1)
                    n += 1

        depth_first_seach(tuple(board), tuple(sorted(list(hand))), 0)

        return self.min_moves if self.min_moves < float("inf") else -1

zuma = input().split(" ")
ans = Solution()
print(ans.findMinStep(zuma[0], zuma[1]))
